File: src/BinaryTreePaths.java
import java.util.ArrayList;
import java.util.List;

public class BinaryTreePaths {

    public static List<String> binaryTreePaths(TreeNode root) {
        List<String> result = new ArrayList<>();
        binaryTreePaths(root, new ArrayList<>(), result);
        return result;
    }

    private static void binaryTreePaths(TreeNode root, List<CharSequence> path, List<String> paths) {
        if (root == null) {
            return;
        }

        path.add(root.val + "");

        if (isLeafNode(root)) {
            paths.add(String.join("->", path));
            path.remove(path.size() - 1);
            return;
        }

        binaryTreePaths(root.left, path, paths);
        binaryTreePaths(root.right, path, paths);
        path.remove(path.size() - 1);
    }

    private static boolean isLeafNode(TreeNode root) {
        return root.left == null && root.right == null;
    }
}

File: python/binary_tree_paths.py
from typing import List


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def is_leaf_node(self, root: TreeNode) -> bool:
        return root.left is None and root.right is None

    def binary_tree_paths(self, root: TreeNode, current_path: str, results: List[str]):
        if root is None:
            return
        if self.is_leaf_node(root):
            results.append(f'{current_path}->{root.val}'[2:])
            return
        self.binary_tree_paths(root.left, f'{current_path}->{root.val}', results)
        self.binary_tree_paths(root.right, f'{current_path}->{root.val}', results)

    def binaryTreePaths(self, root: TreeNode) -> List[str]:
        result = []
        self.binary_tree_paths(root, '', results=result)
        return result
